src/AspNet.Security.OAuth.Untappd/UntappdAuthenticationHandler.cs

```diff
@@ -29,6 +33,63 @@ public class UntappdAuthenticationHandler : OAuthHandler<UntappdAuthenticationOp
         {
         }
 
+        protected override async Task<OAuthTokenResponse> ExchangeCodeAsync([NotNull] OAuthCodeExchangeContext context)
+        {
+            var tokenRequestParameters = new Dictionary<string, string>()
+            {
+                { "client_id", Options.ClientId },
+                { "redirect_uri", context.RedirectUri },
+                { "client_secret", Options.ClientSecret },
+                { "code", context.Code },
+                { "grant_type", "authorization_code" },
+            };
+
+            // PKCE https://tools.ietf.org/html/rfc7636#section-4.5, see BuildChallengeUrl
+            if (context.Properties.Items.TryGetValue(OAuthConstants.CodeVerifierKey, out var codeVerifier))
+            {
+                tokenRequestParameters.Add(OAuthConstants.CodeVerifierKey, codeVerifier!);
+                context.Properties.Items.Remove(OAuthConstants.CodeVerifierKey);
+            }
+
+            var requestContent = new FormUrlEncodedContent(tokenRequestParameters!);
+
+            var builder = new UriBuilder(Options.TokenEndpoint);
+            builder.Port = -1;
+            var query = HttpUtility.ParseQueryString(builder.Query);
+            query["client_id"] = Options.ClientId;
+            query["redirect_uri"] = context.RedirectUri;
+            query["client_secret"] = Options.ClientSecret;
+            query["code"] = context.Code;
+            builder.Query = query.ToString();
+            string url = builder.ToString();
+
+            using var requestMessage = new HttpRequestMessage(HttpMethod.Get, url);
+            requestMessage.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+            requestMessage.Content = requestContent;
+            requestMessage.Version = Backchannel.DefaultRequestVersion;
+
+            var response = await Backchannel.SendAsync(requestMessage, Context.RequestAborted);
+            if (response.IsSuccessStatusCode)
+            {
+                var payload = JsonDocument.Parse(await response.Content.ReadAsStringAsync(Context.RequestAborted));
+                return OAuthTokenResponse.Success(payload);
+            }
+            else
+            {
+                var error = "OAuth token endpoint failure: " + await Display(response);
+                return OAuthTokenResponse.Failed(new Exception(error));
+            }
+        }
+
+        private static async Task<string> Display(HttpResponseMessage response)
+        {
+            var output = new StringBuilder();
+            output.Append("Status: " + response.StatusCode + ";");
+            output.Append("Headers: " + response.Headers.ToString() + ";");
+            output.Append("Body: " + await response.Content.ReadAsStringAsync() + ";");
+            return output.ToString();
```

Comment:
I think it's not a good habit to write like this,  even if this method only handles the error situation, since the underlying library should pay as much attention to performance as possible. 

```C#

string rspText = await response.Content.ReadAsStringAsync();
var output = new StringBuilder(64) // give a suitable capacity
    .Append("Status: ")
    .Append(response.StatusCode) // use specific overload methods
    .Append(';') // use Append(char c)
    .....

```

Reply:
If we were to keep this specific approach I'd agree - this isn't using the full benefits of `StringBuilder` the way you've written it. However, this method should be removed and the error handling should follow the approach used elsewhere for consistency.

Reply:
Go tell the the ASP.Net Core Team, i copied both Methods from there ;-)
https://github.com/dotnet/aspnetcore/blob/a24aed7e5459acc1417a0e490a80e0472d498b2b/src/Security/Authentication/OAuth/src/OAuthHandler.cs

Reply:
Going to replace Error handling as @martincostello suggested to keep it similar.